m3/src/reporter_impl.cc

```diff
@@ -0,0 +1,438 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#include "m3/src/reporter_impl.h"
+
+#include <algorithm>
+#include <cmath>
+#include <iostream>
+#include <limits>
+#include <sstream>
+
+#include "boost/format.hpp"
+#include "thrift/protocol/TCompactProtocol.h"
+
+#include "m3/reporter.h"
+#include "m3/src/calc_transport.h"
+#include "m3/thrift/M3.h"
+#include "tally/src/capable_of.h"
+
+using apache::thrift::protocol::TCompactProtocol;
+using apache::thrift::transport::TTransportException;
+
+namespace tally {
+
+namespace m3 {
+
+const char Reporter::Impl::HISTOGRAM_BUCKET_NAME[] = "bucket";
+const char Reporter::Impl::HISTOGRAM_BUCKET_ID_NAME[] = "bucketid";
```

Comment:
Should these be just `const std::string`?

Reply:
The linter I was using recommended using a pointer to a char array instead of a string since the former doesn't require any dynamic allocation. I found this [SO post](https://stackoverflow.com/questions/47908201/static-global-variable-not-permitted-in-c) discussing the issue a little more. @mway pointed out that since these are passed to the constructor initializer they will create a temporary variable anyway so we might as well use a string. I'll update.

Reply:
I really don't think it matters for constants.

Reply:
For linter I highly recommend clang-tidy.